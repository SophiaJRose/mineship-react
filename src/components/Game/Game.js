import * as React from "react"
import { Board } from "../Board"
import { NewGameButton } from "../NewGameButton/NewGameButton"

function initGame() {
	// Generate unrevealed tiles and tile Images
	let tiles = Array.from({length: 8}, () => Array.from({length: 8}, () => "tile_0"))
	let tileImages = Array.from({length: 8}, () => Array.from({length: 8}, () => "tile_unrevealed"))
	// Generate ships
	let directions = ["u", "r", "d", "l"]
	for (let i = 0; i < 3; i++) {
		// Direction ship is pointing: Up is 0, Right is 1, Down is 2, Left is 3
		let shipDirection = directions[Math.floor(Math.random() * 4)]
		let shipAddPieces = i + 1
		let newShip = []
		let spacesEmpty = true
		// Generate ship positions until empty spots found
		do {
			spacesEmpty = true
			newShip = [[Math.floor(Math.random() * 8), Math.floor(Math.random() * 8)]]
			switch(shipDirection) {
				case "u":
					// Bounds check, if fails, continues the do-while loop
					if (newShip[0][0] + shipAddPieces > 7) {
						spacesEmpty = false
						continue;
					}
					// Ship points up, so next piece increments y-axis
					for (let j = 0; j < shipAddPieces; j++) {
						newShip.push([newShip[j][0] + 1, newShip[j][1]])
					}
					break;
				case "r":
					// Bounds check, if fails, continues the do-while loop
					if (newShip[0][1] - shipAddPieces < 0) {
						spacesEmpty = false
						continue;
					}
					// Ship points right, so next piece decrements x-axis
					for (let j = 0; j < shipAddPieces; j++) {
						newShip.push([newShip[j][0], newShip[j][1] - 1])
					}
					break;
				case "d":
					// Bounds check, if fails, continues the do-while loop
					if (newShip[0][0] - shipAddPieces < 0) {
						spacesEmpty = false
						continue;
					}
					// Ship points down, so next piece decrements y-axis
					for (let j = 0; j < shipAddPieces; j++) {
						newShip.push([newShip[j][0] - 1, newShip[j][1]])
					}
					break;
				case "l":
					// Bounds check, if fails, continues the do-while loop
					if (newShip[0][1] + shipAddPieces > 7) {
						spacesEmpty = false
						continue;
					}
					// Ship points up, so next piece increments x-axis
					for (let j = 0; j < shipAddPieces; j++) {
						newShip.push([newShip[j][0], newShip[j][1] + 1])
					}
					break;
				default:
					// Only here to prevent warning, should be impossible to reach
					console.log("Ship Direction default branch reached")
			}
			// Check none of ship piece locations already filled
			for (let newShipPiece of newShip) {
				if (tiles[newShipPiece[0]][newShipPiece[1]] !== "tile_0") {
					spacesEmpty = false
					break;
				}
			}
		} while (!spacesEmpty)
		// Place ship in tiles
		for (let k = 0; k < newShip.length; k++) {
			// First ship piece is front, last is end, in-between pieces are middle
			switch (k) {
				case 0:
					tiles[newShip[k][0]][newShip[k][1]] = "tile_ship_front_" + shipDirection
					break;
				case newShip.length - 1:
					tiles[newShip[k][0]][newShip[k][1]] = "tile_ship_end_" + shipDirection
					break;
				default:
					tiles[newShip[k][0]][newShip[k][1]] = "tile_ship_middle_" + shipDirection
			}
		}
	}
	// Generate 10 mines
	for (let i = 0; i < 10; i++) {
		let spaceEmpty = true
		let newMine = []
		// Generate mine position until empty position found
		do {
			spaceEmpty = true
			newMine = [Math.floor(Math.random() * 8), Math.floor(Math.random() * 8)]
			if (tiles[newMine[0]][newMine[1]] !== "tile_0") {
				spaceEmpty = false
			}
		} while (!spaceEmpty)
		// Place mine in tiles
		tiles[newMine[0]][newMine[1]] = "tile_mine"
	}
	// Generate numbers
	for (let i = 0; i < 8; i++) {
		for (let j = 0; j < 8; j++) {
			// Skips mines and ship pieces
			if (tiles[i][j] !== "tile_0") {
				continue
			}
			// Get surrounding tiles, with bounds check
			let surroundArea = tiles.map(row => row.slice(Math.max(0,j-1), Math.min(8,j+2))).slice(Math.max(0,i-1), Math.min(8,i+2))
			let surroundTiles = surroundArea.reduce((cumul, arr) => cumul.concat(arr))
			// Count pieces and check if any ship pieces
			let numPieces = 0
			let blue = false
			for (let tile of surroundTiles) {
				if (tile === "tile_mine") {
					numPieces++
				} else if (tile.startsWith("tile_ship")) {
					numPieces++
					blue = true
				}
			}
			// Update, if not 0
			if (numPieces > 0) {
				tiles[i][j] = "tile_" + numPieces + (blue ? "b" : "r")
			}
		}
	}
	// console.log(tiles)
	return {tileImages: tileImages, tiles: tiles}
}

class Game extends React.Component {
	constructor(props) {
		super(props)
		this.state = initGame()
	}

	handleClick(i, j) {
		// Nothing to do if already revealed
		if (this.state.tileImages[i][j] !== "tile_unrevealed") {
			return
		}
		let tileImages = this.state.tileImages.slice()
		tileImages[i][j] = this.state.tiles[i][j]
		this.setState({tileImages: tileImages, tiles: this.state.tiles})
	}

	clickAll() {
		for (let i = 0; i < 8; i++) {
			for (let j = 0; j < 8; j++) {
				this.handleClick(i,j)
			}
		}
	}

	render() {
		return (
			<div>
				<Board onClick={(i, j) => this.handleClick(i, j)} tileImages={this.state.tileImages} />
				<NewGameButton onClick={() => this.setState(initGame())} />
				<button onClick={() => this.clickAll()}>Clear All</button>
			</div>
		);
	}
}

export { Game }