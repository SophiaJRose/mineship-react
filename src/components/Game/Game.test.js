import React from "react"
import { render, fireEvent, screen } from "@testing-library/react"
import "@testing-library/jest-dom/extend-expect"
import { Game } from "./Game"
import tile_unrevealed from "../../images/tile_unrevealed.png"

describe("<Game /> tests", () => {
	it("should render a board", () => {
		const { container } = render(<Game />)
		expect(container.firstChild).toMatchSnapshot()
	})

	it("should change image of tile when clicked", () => {
		// Arrange
		const { container } = render(<Game />)
		// Act
		let tile = container.firstChild.firstChild.firstChild.firstChild
		fireEvent.click(tile)
		// Assert
		expect(tile).toHaveAttribute("src")
		expect(tile).not.toHaveAttribute("src", tile_unrevealed)
	})

	it("should contain 10 mines", () => {
		// Arrange
		const { container } = render(<Game />)
		// Act
		container.firstChild.firstChild.childNodes.forEach((tile) => fireEvent.click(tile))
		// Assert
		expect(screen.getAllByAltText("Mine").length).toBe(10)
	})

	it("should reset after pressing New Game", () => {
		// Arrange
		const { container } = render(<Game />)
		// Act
		fireEvent.click(container.firstChild.firstChild.firstChild)
		let numTilesBefore = screen.getAllByAltText("Unrevealed").length
		fireEvent.click(screen.getByRole("button", {name: /new game/i}))
		let numTilesAfter = screen.getAllByAltText("Unrevealed").length
		// Assert
		expect(numTilesAfter).not.toBe(numTilesBefore)
	})
})