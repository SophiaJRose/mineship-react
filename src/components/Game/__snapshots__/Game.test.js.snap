// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Game /> tests should render a board 1`] = `
<div>
  <h1
    class="title"
  >
    MineShip
  </h1>
  <div
    class="board"
    role="grid"
  >
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
    <button
      class="tile"
    >
      <img
        alt="tile_unrevealed"
        src="tile_unrevealed.png"
      />
    </button>
  </div>
  <div
    class="icons"
  >
    <img
      alt="icon_ship_2.png"
      src="icon_ship_2.png"
    />
    <img
      alt="icon_ship_3.png"
      src="icon_ship_3.png"
    />
    <img
      alt="icon_ship_4.png"
      src="icon_ship_4.png"
    />
  </div>
  <div
    class="menu"
  >
    <button
      class="button"
    >
      New Game
    </button>
    <button
      class="button"
    >
      Rules
    </button>
  </div>
  <span
    aria-label=""
    role="presentation"
  />
  <div
    class="modal"
    id="rulesModal"
  >
    <div
      class="modalContent"
      role="dialog"
    >
      <span
        class="close"
      >
        Ã—
      </span>
      <div>
        <h3>
          Rules
        </h3>
        <p>
          This game is inspired by Minesweeper and Battleship. The goal is to find all of the ships, without finding any of the mines. There are three ships, one that is 2 tiles long, one 3 tiles long and one 4 tiles long.
        </p>
        <p>
          When you click a tile, what's underneath will be revealed. This could be a segment of a ship, a mine, or a number. When a number is revealed, this tells you how many of the 8 surrounding tiles contain either a ship segment, or a mine. If a number is red, that means all the surrounding non-number tiles are mines. If it is blue, at least one of teh surrounding non-number tiles is a ship segment. For example, a red 3 means there are 3 mines in the surrounding tiles. A blue 3 could mean 2 mines and 1 ship segment, 1 mine and 2 ship segments, or 3 ship segments.
        </p>
        <div
          class="rulesImages"
        >
          <img
            alt="3 mines surround the middle tile, so it is a red 3."
            class="rulesImage"
            src="rules_3r.png"
          />
          <img
            alt="2 mines and 1 ship piece produces a blue 3."
            class="rulesImage"
            src="rules_3b_1ship.png"
          />
          <img
            alt="1 mine and 2 ship pieces also produces a blue 3. Note the ship pieces aren't necessarily from the same ship."
            class="rulesImage"
            src="rules_3b_2ship.png"
          />
          <img
            alt="If you're lucky, a blue 3 could mean 3 ship pieces, and 0 mines."
            class="rulesImage"
            src="rules_3b_3ship.png"
          />
        </div>
        <p>
          If you manage to reveal all of the ship segments without revealing any mines, you win the game. If you're unlucky and reveal even a single mine, you'll lose. Whnever the game ends, win or lose, the rest of the board is revealed.
        </p>
      </div>
    </div>
  </div>
</div>
`;
